File: 백준/Gold/1922. 네트워크 연결/네트워크 연결.py
# 알고리즘 : MST, kruskal(parent, union, find, edges, sort, cycle)
# 자료구조 : union-find

import sys
input = sys.stdin.readline

# find 연산 (경로 압축)
def find(x) :
    if parent[x] != x :
        parent[x] = find(parent[x])
    return parent[x]

# union 연산
def union(a, b) :
    a = find(a)
    b = find(b)
    
    if a < b :
        parent[b] = a
    else :
        parent[a] = b

n = int(input())
m = int(input())

# 부모 테이블 정의
parent = [v for v in range(n + 1)]

# 간선 테이블 정의
edges = []

for _ in range(m) :
    a, b, c = map(int, input().split())
    edges.append((c, a, b)) # (비용, a, b) 순으로 저장
    
# 간선 오름차순 정렬
edges.sort()

result = 0 # 총 비용
cnt = 0 # 간선 개수

for edge in edges :
    cost, a, b = edge
    
    if cnt >= n - 1 :
        break
    
    # 사이클이 발생하지 않았다면
    if find(a) != find(b) :
        union(a, b)
        result += cost
        cnt += 1

print(result)

File: 백준/Gold/1922. 네트워크 연결/네트워크 연결.cc
/*
아이디어 : 
알고리즘 : 크루스칼 (parent, edges, sort, cycle)
자료구조 : union-find
*/

#include <bits/stdc++.h>

#define MAX 1001

using namespace std;

int n, m;
int parent[MAX];
vector<pair<int, pair<int, int> > > edges;
int result = 0;

// find 연산 정의 (경로 압축)
int findParent(int x) {
	// 루트 노드가 아니라면, 루트 노드가 나올 때까지 재귀 호출
	if (parent[x] != x)
		parent[x] = findParent(parent[x]);
	return parent[x];
}

void unionParent(int a, int b) {
	a = findParent(a);
	b = findParent(b);
	
	if (a < b) parent[b] = a;
	else parent[a] = b;
}

int main(void) {
	ios::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);
	
	cin >> n >> m;
	
	// 부모 테이블 초기화
	for (int i = 1; i < n + 1; i++) 
		parent[i] = i;
	
	// 간선 정보 입력
	for (int i = 0; i < m; i++) {
		int a, b, cost;
		cin >> a >> b >> cost;
		
		edges.push_back({cost, {a, b}});
	}
	
	// 간선 오름차순 정렬
	sort(edges.begin(), edges.end());
	
	for (int i = 0; i < m; i++) {
		int cost = edges[i].first;
		int a = edges[i].second.first;
		int b = edges[i].second.second;
		
		// 사이클이 생기지 않았다면
		if (findParent(a) != findParent(b)) {
			unionParent(a, b);
			result += cost;
		}
		
		
	}
	
	cout << result << '\n';
	return 0;
}